### PR TITLE
[WIP] peopleforce exporter

### PR DESCRIPTION
**Pull request by pavel-zhuravlyov**

### CODE REVIEW

File: app/services/peopleforce_exporter.rb [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+        {name: 'currency_code', value: Proc.new { |person| nil }}
+    ]
+
+    @employees_table = []
+    @jobs_table = []
+    @dependents_table = []
+    @candidates_table = []
+    @dayoffs_table = []
+
+    Person.all.each do |person|
```
Comment (rilian) [inline app/services/peopleforce_exporter.rb:138]: exclude deleted

Comment (rilian) [reply app/services/peopleforce_exporter.rb:138]: employees are Person with employee hiring_status, Candidates the same with other statuses ..

File: app/services/peopleforce_exporter.rb [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    CSV.open(file, "wb") do |csv|
+      csv << mappings.map { |mapping| mapping[:name] }
+      table.each { |row| csv << row }
+    end
+  end
+
+  EMPLOYEES_MAPPING = [
+      {name: 'employee_id', value: Proc.new { |person| person.employee_id }},
+      {name: 'first_name', value: Proc.new { |person| person.name.split(' ').first }},
+      {name: 'last_name', value: Proc.new { |person| person.name.split(' ').drop(1).join('') }},
```
Comment (rilian) [inline app/services/peopleforce_exporter.rb:61]: `join(' ')` back

File: app/services/peopleforce_exporter.rb [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+      {name: 'email', value: Proc.new { |person| person.email }},
+      {name: 'personal_email', value: Proc.new { |person| person.personal_email }},
+      {name: 'date_of_birth', value: Proc.new { |person| person.day_of_birth }},
+      {name: 'gender', value: Proc.new { |person| nil }},
```
Comment (rilian) [inline app/services/peopleforce_exporter.rb:65]: this is politically incorrect and must be excluded :)

File: app/services/peopleforce_exporter.rb [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+      {name: 'facebook_url', value: Proc.new { |person| nil }},
+      {name: 'linkedin_url', value: Proc.new { |person| person.linkedin }},
+      {name: 'skills', value: Proc.new { |person| person.skills }}
+  ]
+
+  JOBS_MAPPING = [
+      {name: 'employee_id', value: Proc.new { |person| person.employee_id }},
+      {name: 'effective_from', value: Proc.new { |person| person.start_date }},
+      {name: 'compensation', value: Proc.new { |person| nil }},
+      {name: 'employment_type', value: Proc.new { |person| nil }},
```
Comment (rilian) [inline app/services/peopleforce_exporter.rb:80]: if that is employee/contractor then you can get value from `status`

File: app/services/peopleforce_exporter.rb [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+      {name: 'probation_policy', value: Proc.new { |person| nil }},
+      {name: 'position', value: Proc.new { |person| person.current_position }},
+      {name: 'department', value: Proc.new { |person| nil }},
+      {name: 'division', value: Proc.new { |person| nil }},
+      {name: 'location', value: Proc.new { |person| nil }},
```
Comment (rilian) [inline app/services/peopleforce_exporter.rb:85]: person.city

File: app/services/peopleforce_exporter.rb [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+      {name: 'reporting_to_employee_id', value: Proc.new { |person| nil }}
+  ]
+
+  DAYOFFS_MAPPING = [
+      {name: 'employee_id', value: Proc.new { |dayoff_day| dayoff_day.employee_id }},
```
Comment (rilian) [inline app/services/peopleforce_exporter.rb:90]: `dayoff_day.person_id`

File: app/services/peopleforce_exporter.rb [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+      {name: 'date', value: Proc.new { |dayoff_day| dayoff_day.date }},
```
Comment (rilian) [inline app/services/peopleforce_exporter.rb:91]: start_on
https://github.com/rilian/hrms/blob/master/db/structure.sql#L121

File: app/services/peopleforce_exporter.rb (hunk 6 above)
Comment (rilian) [reply app/services/peopleforce_exporter.rb:90]: or dayoff_day.person.employee_id

File: app/services/peopleforce_exporter.rb [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+      {name: 'type', value: Proc.new { |dayoff_day| dayoff_day.type }},
+  ]
+
+  DEPENDENTS_MAPPING = [
```
Comment (rilian) [inline app/services/peopleforce_exporter.rb:95]: we dont have this in HRMS so just remove

File: app/services/peopleforce_exporter.rb (hunk 7 above)
Comment (pavel-zhuravlyov) [reply app/services/peopleforce_exporter.rb:91]: https://github.com/rilian/hrms/pull/128/files/f52bf336525e9c2f97b7e92e38d496a7df2c16c7#diff-16cb3253fe0fc7b6496c0b1fdbb4c9a9R45

**Comment by pavel-zhuravlyov** [comment]
@rilian please review

File: app/services/peopleforce_exporter.rb [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+    @dependents_table = []
+    @candidates_table = []
+    @dayoffs_table = []
+  end
+
+  def perform
+    Person.where(status: Person::CURRENT_EMPLOYEE_STATUSES).each do |person|
+      @employees_table.append(EMPLOYEES_MAPPING.map { |mapping| mapping[:value].call(person).to_s })
+      @jobs_table.append(JOBS_MAPPING.map { |mapping| mapping[:value].call(person).to_s })
+      @dependents_table.append(DEPENDENTS_MAPPING.map { |mapping| mapping[:value].call(person).to_s })
```
Comment (rilian) [inline app/services/peopleforce_exporter.rb:21]: we do not have this data in HRMS, i recommend removing this code

File: app/services/peopleforce_exporter.rb [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+      CANDIDATES_MAPPING.map { |mapping| mapping[:value].call(person).to_s }
+    end
+
+    @dayoffs_table = Dayoff.includes(:person).all
+                         .flat_map { |dayoff| dayoff_to_dayoff_days(dayoff) }
+                         .map { |dayoff_day| DAYOFFS_MAPPING.map { |mapping| mapping[:value].call(dayoff_day).to_s } }
+
+    csv('employees.csv', EMPLOYEES_MAPPING, @employees_table)
+    csv('jobs.csv', JOBS_MAPPING, @jobs_table)
+    csv('dependents.csv', DEPENDENTS_MAPPING, @dependents_table)
```
Comment (rilian) [inline app/services/peopleforce_exporter.rb:34]: same

Comment (pavel-zhuravlyov) [reply app/services/peopleforce_exporter.rb:32]: done

File: app/services/peopleforce_exporter.rb (hunk 9 above)
Comment (pavel-zhuravlyov) [reply app/services/peopleforce_exporter.rb]: done

**Pull request merged**